src/common/procedure/src/procedure.rs

```diff
@@ -0,0 +1,242 @@
+            Status::Done => false,
+        }
+    }
+}
+
+/// Procedure execution context.
+#[derive(Debug)]
+pub struct Context {
+    /// Id of the procedure.
+    pub procedure_id: ProcedureId,
+}
+
+/// A `Procedure` represents an operation or a set of operations to be performed step-by-step.
+#[async_trait]
+pub trait Procedure: Send + Sync {
+    /// Type name of the procedure.
+    fn type_name(&self) -> &str;
+
+    /// Execute the procedure.
+    ///
+    /// The implementation must be idempotent.
+    async fn execute(&mut self, ctx: &Context) -> Result<Status>;
+
+    /// Dump the state of the procedure to a string.
+    fn dump(&self) -> Result<String>;
+
+    /// Returns the [LockKey] if this procedure needs to acquire lock.
+    fn lock_key(&self) -> Option<LockKey>;
+}
+
+/// A key to identify the lock.
+// We might hold multiple keys in this struct. When there are multiple keys, we need to sort the
+// keys lock all the keys in order to avoid dead lock.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LockKey(String);
+
+impl LockKey {
+    /// Returns a new [LockKey].
+    pub fn new(key: impl Into<String>) -> LockKey {
+        LockKey(key.into())
+    }
+
+    /// Returns the lock key.
+    pub fn key(&self) -> &str {
+        &self.0
+    }
+}
+
+/// Boxed [Procedure].
+pub type BoxedProcedure = Box<dyn Procedure>;
+
+/// A procedure with specific id.
+pub struct ProcedureWithId {
+    /// Id of the procedure.
+    pub id: ProcedureId,
+    pub procedure: BoxedProcedure,
+}
+
+impl ProcedureWithId {
+    /// Returns a new [ProcedureWithId] that holds specific `procedure`
+    /// and a random [ProcedureId].
+    pub fn with_random_id(procedure: BoxedProcedure) -> ProcedureWithId {
+        ProcedureWithId {
+            id: ProcedureId::random(),
+            procedure,
+        }
+    }
+}
+
+/// Unique id for [Procedure].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct ProcedureId(Uuid);
+
+impl ProcedureId {
+    /// Returns a new unique [ProcedureId] randomly.
+    pub fn random() -> ProcedureId {
+        ProcedureId(Uuid::new_v4())
+    }
+
+    /// Parses id from string.
+    pub fn parse_str(input: &str) -> Result<ProcedureId> {
+        Uuid::parse_str(input)
+            .map(ProcedureId)
+            .context(InvalidIdSnafu)
+    }
+}
+
+impl fmt::Display for ProcedureId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Loader to recover the [Procedure] instance from serialized data.
+pub type BoxedProcedureLoader = Box<dyn Fn(&str) -> Result<BoxedProcedure> + Send>;
+
+// TODO(yingwen): Find a way to return the error message if the procedure is failed.
+/// State of a submitted procedure.
+#[derive(Debug)]
+pub enum ProcedureState {
+    /// The procedure is running.
+    Running,
+    /// The procedure is finished.
+    Done,
+    /// The procedure is failed.
+    Failed,
```

Comment:
Do we need to add a field to indicate if the failure is temporary (can be rerun) or permenant (should not rerun) like [opendal](https://docs.rs/opendal/0.25.1/opendal/struct.Error.html#method.is_temporary) does?

Reply:
The failure should be permenant, I will explain this in the comment.